### ISSUE TITLE
Use a nonce to calculate state hash

### ISSUE THREAD

**Issue by fleupold** [gnosis/dex-services#44]
We currently calculate the account state hash by concatenating all balances into one rolling hash.

```
State([1, 2, 3,4] = H(H(H(H(0, 1), 2), 3), 4)
```

This means, that two states with the same balances (e.g. all accounts owning 0 after an initial deposit + withdraw).

While we cannot identify a concrete attack vector right now, it seems like good practice to avoid such collisions by also adding a changing salt/nonce into the procedure.

Therefore, we should either introduce a stateId or simply the previous state hash as the starting point for the balance hash creation.

```
State_i+1([1, 2, 3,4] = H(H(H(H(State_i, 1), 2), 3), 4)
```

**Comment by bh2smith** [comment]
I like this idea very much and am completely in favour of doing this!

**Comment by fleupold** [comment]
Closed by #120
